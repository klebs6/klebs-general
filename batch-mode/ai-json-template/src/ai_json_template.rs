// ---------------- [ File: ai-json-template/src/ai_json_template.rs ]
crate::ix!();

/// The derived code implements `AiJsonTemplate` for each struct, letting you
/// call `MyStruct::to_template()` to get a JSON “schema” describing how the
/// AI should produce data that matches this layout.
///
pub trait AiJsonTemplate
: Clone 
+ Debug 
+ LoadFromFile 
+ SaveToFile<Error=SaveLoadError> 
+ Serialize 
+ for<'a> Deserialize<'a> 
{
    /// Return a JSON template describing how the AI’s output should be structured.
    /// This might include doc comments or other instructions for each field.
    fn to_template() -> serde_json::Value;
}
