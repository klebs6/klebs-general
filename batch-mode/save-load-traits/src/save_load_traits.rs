// ---------------- [ File: src/save_load_traits.rs ]
crate::ix!();

#[async_trait]
pub trait SaveToFile {

    type Error;

    async fn save_to_file(
        &self,
        filename: impl AsRef<Path> + Send,

    ) -> Result<(), Self::Error>;
}

#[async_trait]
pub trait LoadFromFile: Sized {

    type Error;

    async fn load_from_file(filename: impl AsRef<Path> + Send) 
        -> Result<Self, Self::Error>;

}
