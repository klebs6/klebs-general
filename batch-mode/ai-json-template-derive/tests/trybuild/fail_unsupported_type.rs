// ---------------- [ File: tests/trybuild/fail_unsupported_type.rs ]
use ai_json_template_derive::*;
use ai_json_template::*;
use save_load_traits::*;
use save_load_derive::*;
use batch_mode_3p::*;
use serde::{Serialize, Deserialize};

#[derive(SaveLoad,Debug,Clone,AiJsonTemplate, Serialize, Deserialize)]
struct BadConfig {
    /// i32 is not allowed if your macro only supports String, Option<String>, Vec<String>, or nested.
    number: i32,
}

fn main() {
    // We expect a compile-time error about "Unsupported field type for AiJsonTemplate: i32"
    // The test harness calls `t.compile_fail("...")` on this file.
}
