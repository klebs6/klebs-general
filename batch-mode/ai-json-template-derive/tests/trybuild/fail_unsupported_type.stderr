error: Unsupported field type for AiJsonTemplate: i32
  --> tests/trybuild/fail_unsupported_type.rs:10:13
   |
10 |     number: i32,
   |             ^^^
